tmt/base.py

```diff
@@ -237,16 +237,206 @@ def _validate_fmf_node(self, node: fmf.Tree, logger: tmt.utils.Common,
+    Mixin adding support for loading fmf keys into object attributes.
+
+    When invoked, annotated class-level variables are searched for in a given fmf node,
+    and if the key of the same name as the variable exists, its value is "promoted" to
+    instance variable.
+
+    If a method named ``_normalize_<variable name>`` exists, it is called with the fmf
+    key value as its single argument, and its return value is assigned to instance
+    variable. This gives class chance to modify or transform the original value when
+    needed, e.g. to convert the original value to a type more suitable for internal
+    processing.
+    """
+
+    # If specified, keys would be iterated over in the order as listed here.
+    KEYS_SHOW_ORDER: List[str] = []
+
+    # NOTE: these could be static methods, self is probably useless, but that would
+    # cause complications when classes assign these to their members. That makes them
+    # no longer static as far as class is concerned, which means they get called with
+    # `self` as the first argument. A workaround would be to assign staticmethod()-ized
+    # version of them, but that's too much repetition.
+    def _normalize_string_list(self, value: Union[None, str, List[str]]) -> List[str]:
+        if value is None:
+            return []
+
+        return [value] if isinstance(value, str) else value
+
+    def _normalize_environment(self, value: Optional[Dict[str, Any]]) -> tmt.utils.EnvironmentType:
+        if value is None:
+            return {}
+
+        return {
+            name: str(value) for name, value in value.items()
+            }
+
+    @classmethod
+    def _iter_key_annotations(cls) -> Generator[Tuple[str, Any], None, None]:
+        """
+        Iterate over keys' type annotations.
+
+        Keys are yielded in the order: keys declared by parent classes first, then
+        keys declared by the class itself, all following the order in which keys
+        were defined in their respective classes.
+
+        Yields:
+            pairs of key name and its annotations.
+        """
+
+        for base in cls.__bases__:
+            for name, value in base.__dict__.get('__annotations__', {}).items():
+                # Skip special fields that are not keys.
+                if name == 'KEYS_SHOW_ORDER':
+                    continue
+
+                yield (name, value)
+
+        yield from cls.__dict__.get('__annotations__', {}).items()
+
+    @classmethod
+    def iter_key_names(cls) -> Generator[str, None, None]:
+        """
+        Iterate over key names.
+
+        Keys are yielded in the order: keys declared by parent classes first, then
+        keys declared by the class itself, all following the order in which keys
+        were defined in their respective classes.
+
+        Yields:
+            key names.
+        """
+
+        for keyname, _ in cls._iter_key_annotations():
+            yield keyname
+
+    def iter_keys(self) -> Generator[Tuple[str, Any], None, None]:
```

Comment:
Name a bit misleading as it does not yield only keys. What about `iter_keys_names` or `iter_items`? Or consistently with `SerializableContainer` use `keys()` and `items()`?

Reply:
Hm, that's probably it, this PR was born long before `SerializableContainer`, and I recall you suggested this change for `SerializableContainer` - made it there, forgot to stick to the naming scheme here. I'll check what can be done, but I think there's no overlap with existing methods of `Core` & children.

Reply:
Done.